### PR TITLE
Add command to migrate legacy dictionaries to WebExtension format

### LINKED ISSUE #8587: Migrate Dictionaries to Web Extension format
Automatically convert all dictionaries on AMO to the new format. This can't be done until they are supported on release, and will require notifying developers as well.

### PR DESCRIPTION
**Pull request by diox**
Fix #8587

### CODE REVIEW

File: src/olympia/addons/tests/test_tasks.py [hunk 1]
```diff
@@ -249,3 +252,64 @@ def test_add_static_theme_broken_lwt(self):
...
+@override_settings(ENABLE_ADDON_SIGNING=True)
+class TestMigrateLegacyDictionaryToWebextension(TestCase):
+    def setUp(self):
+        self.user = user_factory(
+            id=settings.TASK_USER_ID, username='taskuser',
+            email='taskuser@mozilla.com')
+        with freeze_time('2017-07-27 07:00'):
+            self.addon = addon_factory(
+                type=amo.ADDON_DICT, target_locale='fr',
+                version_kw={'version': '6.3'})
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_tasks.py]: After #8068 `addon_factory()` defaults to a random created date (or rather, it was already supposed to be a random date but it was broken before) so freeze_time doesn't do anything.

File: src/olympia/addons/tests/test_utils_.py [hunk 2]
```diff
@@ -278,3 +278,66 @@ def test_lwt_missing_info(self):
...
+        with zipfile.ZipFile(xpi_file_path, 'r', zipfile.ZIP_DEFLATED) as xpi:
+            # Check that manifest is present, contains proper version and
+            # dictionaries properties.
+            manifest = xpi.read('manifest.json')
+            manifest_json = json.loads(manifest)
+            assert manifest_json['version'] == expected_version
+            expected_dict_obj = {addon.target_locale: 'dictionaries/ar.dic'}
+            assert manifest_json['dictionaries'] == expected_dict_obj
+
+            # Check that we haven't included any useless files.
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_utils_.py]: where do we add useless files?

Comment (diox) [reply src/olympia/addons/tests/test_utils_.py]: The file we're testing with (`src/olympia/files/fixtures/files/dictionary-test.xpi`) already has an `install.js` file as well as a `__MACOSX` directory. And obviously, an `install.rdf` file.

File: src/olympia/addons/tests/test_tasks.py (hunk 1 above)
Comment (diox) [reply src/olympia/addons/tests/test_tasks.py]: Not the version though, and that's what I cared about (for `current_version` "promotion" to work, I need the older version to be guaranteed to be in the past compared to the new one I'm creating)

File: src/olympia/addons/management/commands/process_addons.py [hunk 3]
```diff
@@ -77,7 +79,19 @@
...
+        ]
+    },
+    'migrate_legacy_dictionaries_to_webextension': {
+        'method': migrate_legacy_dictionaries_to_webextension,
+        'qs': [
+            Q(type=amo.ADDON_DICT,
+              status=amo.STATUS_PUBLIC,
+              disabled_by_user=False,
+              _current_version__files__is_webextension=False,
+              target_locale__isnull=False),
```
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/addons/management/commands/process_addons.py]: Well, I would have loved to support the others but:
- I need the locale in the new manifest, but it's nowhere in the old one. It's not in any of the files in the legacy .xpi, in fact!
- They are really hard to find for end-users anyway

File: src/olympia/addons/management/commands/process_addons.py [hunk 4]
```diff
@@ -77,7 +79,19 @@
...
               _current_version__files__is_webextension=True)
-        ]},
+        ]
+    },
+    'migrate_legacy_dictionaries_to_webextension': {
+        'method': migrate_legacy_dictionaries_to_webextension,
+        'qs': [
+            Q(type=amo.ADDON_DICT,
+              status=amo.STATUS_PUBLIC,
+              disabled_by_user=False,
```
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/addons/management/commands/process_addons.py]: I was thinking they'd be forcibly disabled by admins when we run the migration. I could include them, just didn't want to run into different file path issues...

File: src/olympia/addons/management/commands/process_addons.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/addons/management/commands/process_addons.py]: > They are really hard to find for end-users anyway

You're right about that - new frontend doesn't expose any link to browse all dictionaries and doesn't even list "dictionary" as a supported type on the search page.  (https://addons.mozilla.org/en-GB/firefox/search/?type=dictionary works regardless).  The issue _did_ say "all dictionaries" though.  

@jvillalobos what's the plan with dictionaries?  Is it okay that we will only be converting dictionaries listed on the language tools page?  (Should they be exposed as a type in frontend at least, etc?)

> It's not in any of the files in the legacy .xpi, in fact!

It's the file name of the `.dic` file, isn't it?

[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/addons/management/commands/process_addons.py]: > It's the file name of the .dic file, isn't it?

Not necessarily, no - AFAIK there is no guarantee on that. For instance in the french dictionary, the filename is `dictionaries/fr-classic.dic`, which doesn't map to any locale.

[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/addons/management/commands/process_addons.py]: > As for converting them, I don't think we should be selective, and just convert all of them. 

we shouldn't be selective so  _we should convert them all_? or we should be selective and _don't convert them all_?

[the remark this answers is not in the document]
Comment (jvillalobos) [reply src/olympia/addons/management/commands/process_addons.py]: Heh, sorry. I think we should convert them all.

[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/addons/management/commands/process_addons.py]: Can you set `target_locale` on all of them then ? I can't programmatically guess it - see my answer above.

**Comment by diox** [comment]
Note that Firefox still supports the legacy dictionaries at the moment anyway. So any non-migrated dictionary can stay that way for a little while. Since the command will pick up any non-disabled legacy dictionaries with a `target_locale` set, we can run it multiple times to pick up any legacy dicts not converted the first time around, once they are properly set up.

File: src/olympia/addons/management/commands/process_addons.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (diox) [reply src/olympia/addons/management/commands/process_addons.py]: Filed https://github.com/mozilla/addons-server/issues/9344 for this edge case

**Pull request merged**